### PR TITLE
Wrap generic sparse-sparse GEMM

### PR DESCRIPTION
**Pull request by kshyatt**

### CODE REVIEW

**Comment by maleadt** [comment]
:tada:

Fixes https://github.com/JuliaGPU/CUDA.jl/issues/1264?

CI failure looks related though; this functionality probably isn't available on 10.x.

**Comment by kshyatt** [comment]
Update: not quite working, I'll update as soon as I have a fix

**Comment by kshyatt** [comment]
After investigating using the CUDA library samples script, I think the underlying CUDA code is bugged for `beta != 0`. So I'm ok with merging this as-is if others are.

**Comment by maleadt** [comment]
What about erroring (or at least warning once) at run time if beta != 0?

**Comment by maleadt** [comment]
LGTM. Let's maybe hold until I hear back from NVIDIA. Please ping me if I would forget about this!

**Comment by frapac** [comment]
Thanks for wrapping `SpGEMM`, this function is working quite well for our use case! 
I only have a generic question regarding the new API: `cuSPARSE` requires to allocate one or two buffers each time we call a function (`SpMV`, `SpGEMM`, `SpSV`, ...), which is fine if we want to compute the operation only once. However, this can prove detrimental if we have to repeat several times the same operation (imagine we want to compute 1,000 `SpGEMM` with `A`, `B` and `C` having always the same sparsity pattern). In our use case, with matrices with size ~ `40000 x 20000`, we already get ~20MB allocations on the GPU per `SpGEMM` call:
```julia
julia> CUDA.@time mm!('N', 'N', 1.0, Jt, J, 0.0, C, 'O')
  0.009050 seconds (93 CPU allocations: 3.094 KiB) (2 GPU allocations: 22.328 MiB, 4.05% memmgmt time)
```
which is just intractable for us (again, this is not related to your implementation, but the way cuSPARSE has been designed).

My question is more regarding the long term: what would be a good API in `cuSPARSE` to cache the buffers (for instance if we want to wrap [`SpGEMMreuse`](https://docs.nvidia.com/cuda/cusparse/index.html#cusparse-generic-function-spgemm-reuse)? I think this is also relevant if we want to wrap the new functions `SpSV` and `SpSM` to solve triangular systems on the GPU.

File: test/cusparse.jl [hunk 1]
```diff
@@ -971,6 +971,32 @@ end
...
+                @testset "$(typeof(d_A))" for (d_A, d_B, d_C) in [(CuSparseMatrixCSR(A),
+                                                                   CuSparseMatrixCSR(B),
+                                                                   CuSparseMatrixCSR(C))] # for now only supports CSR
+                    mm! = CUSPARSE.mm!
+                    @test_throws ArgumentError mm!('N','T',alpha,d_A,d_B,beta,d_C,'O')
+                    @test_throws ArgumentError mm!('T','N',alpha,d_A,d_B,beta,d_C,'O')
+                    @test_throws ArgumentError mm!('T','T',alpha,d_A,d_B,beta,d_C,'O')
+                    @test_throws DimensionMismatch mm!('N','N',alpha,d_A,d_B,beta,d_B,'O')
+                    D = alpha * A * B + beta * C
+                    d_D = mm!('N','N',alpha,d_A,d_B,beta,d_C,'O')
```
Comment (maleadt) [inline test/cusparse.jl]: this isn't really a separate `d_D`, but instead the call mutates `d_C`. A non-mutating version, `mm`, could return a `D` matrix instead.

**Comment by kshyatt** [comment]
Adding this `@sync` seems to have fixed things locally for me.

**Comment by kshyatt** [comment]
@frapac thank you for the feedback about reuse -- sorry it took so long for someone to respond. This is an interesting problem, but it might be better suited to its own Issue for us to discuss. Would you mind opening one?

**Comment by frapac** [comment]
Sure, I have opened a separate issue here: #1362 
And no worry for the delayed answer: on the contrary, we are using extensively CUSPARSE in our current project, and we really appreciate all the works you have done so far for that library!

**Comment by kshyatt** [comment]
Tests passed locally. Now works for any `beta`, any sparsity pattern of `A*B` and `C` thanks to https://github.com/JuliaGPU/CUDA.jl/pull/1380. DO NOT MERGE this until that PR has been merged.

File: lib/cusparse/generic.jl [hunk 2]
```diff
@@ -223,3 +223,74 @@ function mm!(transa::SparseChar, transb::SparseChar, alpha::Number, A::CuSparseM
...
+        C.rowPtr = CuVector{Cint}(undef, Cm[] + 1)
+        C.colVal = CuVector{Cint}(undef, Cnnz1[])
+        C.nzVal  = CuVector{T}(undef, Cnnz1[])
+        C.nnz    = Cnnz1[]
+        cusparseCsrSetPointers(descC, C.rowPtr, C.colVal, C.nzVal)
+        cusparseSpGEMM_copy(handle(), transa, transb, Ref{T}(alpha), descA, descB,
+                            Ref{T}(beta), descC, T, CUSPARSE_SPGEMM_DEFAULT, spgemm_Desc)
+        return C
+    else
+        newC = CuSparseMatrixCSR(CUDA.zeros(Cint, Cm[] + 1), CUDA.zeros(Cint, Cnnz1[]), CUDA.zeros(T, Cnnz1[]), (Cm[], Cn[]))
```
Comment (maleadt) [inline lib/cusparse/generic.jl:291]: Maybe add a note here how SpGEMM with beta != 0 only supports identical sparsity patterns, making it unusable?

**Pull request merged**